asf_heat_pump_suitability/pipeline/reweight_epc/prepare_sample.py

```diff
@@ -0,0 +1,123 @@
+import polars as pl
+import polars.selectors as cs
+from asf_heat_pump_suitability import config
+
+
+def drop_nulls_feature_cols(df, features):
+    """
+    Drop rows with null values in any feature column from EPC dataset.
+
+    Args:
+        df (pl.DataFrame): EPC dataset
+        features (list): column names of features
+
+    Returns:
+        pl.DataFrame: EPC dataset where rows with null in any specified feature column are dropped
+    """
+    df = (
+        df.with_columns(  # replace "unknown" with null in string feature columns
+            (cs.string() & cs.by_name(features)).replace(
+                {"unknown": None}, return_dtype=pl.String
+            )
+        )
+        .with_columns(
+            pl.col("build_year").fill_null("unknown")
+        )  # replace null with unknown because it's a build_year category
```

Comment:
We can keep build year (I think) because the target data has an 'unknown' category, whereas for the other features this doesnt exist so we can't keep unknowns. I think the reason this one has 'unknown' in the target but the others dont is that the target data for build year is from CDRC and the others are all census data. Let me know if you think a different approach is better.

To 2nd q: I checked and actually there are no nulls in the original `CONSTRUCTION_AGE_BAND` column in the EPC dataset we're using from which `build_year` is derived so seems like this line does nothing with the current EPC iteration. Do you think we should drop nulls instead of fill them if there ever are nulls?

Reply:
this all makes sense! I think it's good to keep the weighting pipeline as you have it - it's a shame the others don't have an unknown category in their target datasets. I had forgotten it was due to the target data.

I don't think you should drop nulls if there ever are them... makes sense that these would mean the same thing as 'unknown'?

On removing these rows after the weighting... I can see pros and cons! I think keep them in for now. We could always think about removing these properties later down the line when calculating HP suitability, but they are a minority anyway so kind of think it won't make a difference anyway!

Reply:
Thanks for checking this Liz. I agree that for EPC there's not likely to be a diff between 'unknown' and null values. Now that I think about it, it's possible that in the EPC processing pipeline all nulls get filled with 'unknown' anyway.